viewing_party/party.py

```diff
@@ -1,23 +1,175 @@
+        user_data['watched'].append(movie_to_move)
+    
+    return user_data
+          
 
-# -----------------------------------------
 # ------------- WAVE 2 --------------------
-# -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    watched_movies = user_data.get('watched', [])
+
+    rating_sum = 0.0
+    total_movies = 0
+
+    for movie in watched_movies:
+        rating = movie.get('rating')
+        if rating is not None:
+            rating_sum += rating
+            total_movies += 1
+
+    if total_movies == 0:
+        return 0.0  
+
+    average_rating = rating_sum / total_movies
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+
+    watched_movie = user_data.get('watched',[])
+    
+    #count occurences of each genre
+    genre_count = {}
+    most_watched_genre = ""
+    max_count = 0
+     
+    for movie in watched_movie:
+        genre = movie.get('genre', None)
+
+        if genre:
+            if genre not in genre_count:
+                genre_count[genre] = 1
+            else:           
+                genre_count[genre] += 1
+
+        if genre_count[genre] > max_count:
+            max_count = genre_count[genre]
+            most_watched_genre = genre
+
+    if watched_movie == []:
+        return None
 
-# -----------------------------------------
+    return most_watched_genre
+        
+               
 # ------------- WAVE 3 --------------------
-# -----------------------------------------
+
+def get_unique_watched(user_data):
+    user_unique_watched_movies = []
+    watched_movie = user_data.get('watched',[])
+    friends = user_data.get('friends',[])
+   
+    friends_watched_titles = []
+    for friend in friends:
+        for friend_movie in friend.get('watched', []):
+            friends_watched_titles.append(friend_movie['title'])
+    for movie in watched_movie:
+        if movie['title'] not in friends_watched_titles:
+            user_unique_watched_movies.append(movie)    
+    print(user_unique_watched_movies)
+    return user_unique_watched_movies
+
+
+def get_friends_unique_watched(user_data):
+    friends_unique_watched_movies_list = []
+
+    user_watched_title_set = set()   
+    user_watched_movie_list = user_data.get('watched', [])
+
+    friends_watched_movie_list = user_data.get('friends', [])
+    friends_watched_title_set = set()
+    
+    for user_watched_movies in user_watched_movie_list:
+        user_watched_title_set.add(user_watched_movies['title'])
+
+    for friends_movies_list in friends_watched_movie_list: #list
+        for firends_watched_movie in friends_movies_list.get('watched', []): #dict.value
+            if firends_watched_movie['title'] not in user_watched_title_set and firends_watched_movie['title'] not in friends_watched_title_set:
+                friends_unique_watched_movies_list.append(firends_watched_movie)
+            friends_watched_title_set.add(firends_watched_movie['title'])

```

Comment:
I see here that you are making use of a `set` so that you have a constant look up time, I love that! The only thing I might suggest is that the readability here is not where it could be. What changes to this part of the code could increase readability?